### PR TITLE
Disable deploying global resources by default

### PR DESCRIPTION
**Pull request by dturn**
**What are you trying to accomplish with this PR?**
As part of the 1.0 push we are going to require a new command to deploy global resources https://github.com/Shopify/kubernetes-deploy/issues/522

This is the first step, preventing `krane deploy` from deploying global resources. This does maintain backwards compatibility with kubernetes-deploy. 

**How is this accomplished?**
Fetch global resources from the cluster and warn/raise if any are part of the deploy set.

**What could go wrong?**
We miss global resources or this change upsets people.

### CODE REVIEW

File: lib/kubernetes-deploy/deploy_task.rb [hunk 1]
```diff
@@ -100,13 +101,17 @@ def prune_whitelist
...
+    end
+
     def server_version
       kubectl.server_version
     end
 
     def initialize(namespace:, context:, current_sha:, logger: nil, kubectl_instance: nil, bindings: {},
       max_watch_seconds: nil, selector: nil, template_paths: [], template_dir: nil, protected_namespaces: nil,
-      render_erb: true)
+      render_erb: true, allow_globals: false)
```
Comment (dturn) [inline lib/kubernetes-deploy/deploy_task.rb]: Allows us to maintain backward compatibility of kubernetes-deploy. This will be removed when the exe files go away

Comment (KnVerey) [reply lib/kubernetes-deploy/deploy_task.rb]: 🤔Makes sense, but we should probably make sure that `Krane::DeployTask` never has this option. Maybe the transitional version we release will also need to have two classes in addition to two executables, where the `Krane` one is just delegating to `KubernetesDeploy`. Maybe we release the transitional version as `krane`, but as `0.1.0` and then `1.0.0` is when kubernetes-deploy is ripped out?

Comment (dturn) [reply lib/kubernetes-deploy/deploy_task.rb]: I know rails used to release `X.99.0` right before `X+1.0.0` so you could bump to `X.99.0` and check for deprecation warnings before bumping to `X+1.0.0`. 

We're in a bit of branch from brach crunch. But, that could be a good pattern for making breaking changes.

File: lib/kubernetes-deploy/cluster_resource_discovery.rb [hunk 2]
```diff
@@ -16,8 +16,24 @@ def crds
       end
     end
 
+    def globals
+      @globals ||= fetch_globals.map do |gv|
+        kind, group = gv.split(".", 2)
+        { group: group, kind: kind.singularize }
```
Comment (KnVerey) [inline lib/kubernetes-deploy/cluster_resource_discovery.rb]: `singularize` isn't necessarily going to give us the real kind. You can get it from kubectl if you use `-o wide` instead of `-o name`. Seems like the kind is the only part you actually use anyway, so maybe this should just be `global_kinds` and return an array?

Incidentally, I noticed the docs have a worrying disclaimer: "(This output can be retrieved from kubectl api-resources, and is accurate as of Kubernetes 1.13.3.)" but it [turns out](https://github.com/kubernetes/website/pull/13223) that was supposed to mean that the copy of the output mentioned in the docs was snapshotted from 1.13.3. So this should be fine to use, as you'd hope.

File: lib/krane/deploy_task.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+require 'kubernetes-deploy/deploy_task'
+
+module Krane
+  class DeployTask < KubernetesDeploy::DeployTask
+    def initialize(**args)
```
Comment (KnVerey) [inline lib/krane/deploy_task.rb]: Not sure how much this matters, but technically doesn't this mean `Krane::DeployTask` _does_ have the `allow_globals` argument (you won't get an argument error), but we overwrite it? We could manually raise the argument error I guess.

File: lib/kubernetes-deploy/deploy_task.rb [hunk 4]
```diff
@@ -261,16 +267,35 @@ def validate_resources(resources)
...
 
+    def validate_globals(resources)
+      return unless (global = resources.select(&:global?).presence)
+      global_names = global.map do |resource|
+        "#{resource.name} (#{resource.type}) in #{File.basename(resource.file_path)}"
+      end
+      global_names = FormattedLogger.indent_four(global_names.join("\n"))
+
+      if @allow_globals
+        msg = "The ability for this task to deploy global resources will be removed in the next version:"
```
Comment (KnVerey) [inline lib/kubernetes-deploy/deploy_task.rb]:
```suggestion
        msg = "The ability for this task to deploy global resources will be removed in the next version, which will affect the following resources:"
```

File: lib/kubernetes-deploy/deploy_task.rb [hunk 5]
```diff
@@ -261,16 +267,35 @@ def validate_resources(resources)
...
+        msg += "\n#{global_names}"
+        @logger.summary.add_paragraph(ColorizedString.new(msg).yellow)
+      else
+        @logger.summary.add_paragraph(ColorizedString.new("Global resources:\n#{global_names}").yellow)
+        raise FatalDeploymentError, "Deploying global resource is not allowed from this command."
```
Comment (KnVerey) [inline lib/kubernetes-deploy/deploy_task.rb]:
```suggestion
        raise FatalDeploymentError, "This command is namespaced and cannot be used to deploy global resources."
```

File: lib/kubernetes-deploy/kubernetes_resource.rb [hunk 6]
```diff
@@ -50,8 +50,13 @@ def build(namespace:, context:, definition:, logger:, statsd_tags:, crd: nil)
         if crd
           CustomResource.new(crd: crd, **opts)
         else
-          inst = new(**opts)
-          inst.type = definition["kind"]
+          type = definition["kind"]
+          inst = if globals.include?(type.downcase)
+            KubernetesDeploy::GlobalKubernetesResource.new(**opts)
```
Comment (KnVerey) [inline lib/kubernetes-deploy/kubernetes_resource.rb]: What if instead of having a new class, we make `global` settable, and handle it like we do `type` below? I think might be a bit simpler overall.

File: test/integration/kubernetes_deploy_test.rb [hunk 7]
```diff
@@ -1593,6 +1592,25 @@ def test_fail_erb_templates_if_rendering_is_disabled
     ], in_order: true)
   end
 
+  def test_deploy_allow_globals_warns
+    result = deploy_fixtures("globals")
```
Comment (KnVerey) [inline test/integration/kubernetes_deploy_test.rb]: Don't we already have a fixture set with a global resource in it (CRD or storage class probably?)?

File: test/helpers/fixture_deploy_helper.rb [hunk 8]
```diff
@@ -62,7 +62,7 @@ def deploy_raw_fixtures(set, wait: true, bindings: {}, subset: nil)
 
   def deploy_dirs_without_profiling(dirs, wait: true, allow_protected_ns: false, prune: true, bindings: {},
     sha: "k#{SecureRandom.hex(6)}", kubectl_instance: nil, max_watch_seconds: nil, selector: nil,
-    protected_namespaces: nil, render_erb: true)
+    protected_namespaces: nil, render_erb: true, allow_globals: true)
```
Comment (KnVerey) [inline test/helpers/fixture_deploy_helper.rb]: Is the plan to change the default when we remove `KubernetesDeploy`?

File: lib/krane/deploy_task.rb [hunk 9]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+require 'kubernetes-deploy/deploy_task'
+
+module Krane
+  class DeployTask < KubernetesDeploy::DeployTask
```
Comment (KnVerey) [inline lib/krane/deploy_task.rb]: Does this implicitly get coverage from your CLI tests, or should we be using the krane version of the code a new integration test?

File: lib/kubernetes-deploy/deploy_task.rb (hunk 1 above)
Comment (KnVerey) [reply lib/kubernetes-deploy/deploy_task.rb]: For backwards compatibility, shouldn't the default be true?

File: lib/krane/deploy_task.rb (hunk 9 above)
Comment (dturn) [reply lib/krane/deploy_task.rb]: It gets coverage form the black_box_tests

File: test/helpers/fixture_deploy_helper.rb (hunk 8 above)
Comment (dturn) [reply test/helpers/fixture_deploy_helper.rb]: yep.

File: test/integration/kubernetes_deploy_test.rb (hunk 7 above)
Comment (dturn) [reply test/integration/kubernetes_deploy_test.rb]: We do, my thought was this would be the start of the `hello-cloud` but for global resources. I'm fine to save that for the first person to actually add real code to a global resource.

File: lib/kubernetes-deploy/kubernetes_resource.rb (hunk 6 above)
Comment (dturn) [reply lib/kubernetes-deploy/kubernetes_resource.rb]: I went back and forth on this couldn't decide which was better. This felt more future proof, though I can't think of any special behavior we'd want from a GlobalResource., but the approach of not using a new class was definitely simpler. I'll try the other approach and see  how it goes.

File: lib/kubernetes-deploy/cluster_resource_discovery.rb (hunk 2 above)
Comment (dturn) [reply lib/kubernetes-deploy/cluster_resource_discovery.rb]: I'd prefer to use `singularize` over parsing the output to wide, but 👍 to the array approach.

Comment (KnVerey) [reply lib/kubernetes-deploy/cluster_resource_discovery.rb]: It really isn't safe to assume that `singularize` (some logic somebody wrote into ruby) will generate the real `kind` name (which is explicitly chosen by someone writing a new resource). In theory the plural, the singular and the kind could all be completely different words. I'm not sure there are examples of unexpected plural forms in core resources, but CRs can be global too.

File: lib/kubernetes-deploy/kubernetes_resource.rb (hunk 6 above)
Comment (KnVerey) [reply lib/kubernetes-deploy/kubernetes_resource.rb]: 🤔 If we considered server resource lists part of the task's global operating context and put them into our new task config object, then we'd have a third option that might read even more straightforwardly: inside the normal superclass, we could do essentially `def global?; @task_config.global_resource_kinds.include?(kind); end`.

File: test/integration/kubernetes_deploy_test.rb (hunk 7 above)
Comment (KnVerey) [reply test/integration/kubernetes_deploy_test.rb]: True, we're going to have to separate that out entirely, so a purpose-built fixture set makes sense.

File: lib/kubernetes-deploy/kubernetes_resource.rb (hunk 6 above)
Comment (dturn) [reply lib/kubernetes-deploy/kubernetes_resource.rb]: I think that makes sense to do in https://github.com/Shopify/kubernetes-deploy/issues/572.

File: lib/kubernetes-deploy/cluster_resource_discovery.rb (hunk 2 above)
Comment (dturn) [reply lib/kubernetes-deploy/cluster_resource_discovery.rb]: Good point, code now parses wide. This was a bit tricker than expected, some columns don't have values so just splitting on whitespace doesn't work, instead I took advantage of the fixed width of each column.

File: lib/kubernetes-deploy/deploy_task.rb [hunk 10]
```diff
@@ -100,13 +100,17 @@ def prune_whitelist
       wl + cluster_resource_discoverer.crds.select(&:prunable?).map(&:group_version_kind)
     end
 
+    def global_resource_names
```
Comment (KnVerey) [inline lib/kubernetes-deploy/deploy_task.rb]: why is this method public?

File: lib/kubernetes-deploy/kubernetes_resource.rb [hunk 11]
```diff
@@ -50,8 +50,10 @@ def build(namespace:, context:, definition:, logger:, statsd_tags:, crd: nil)
         if crd
           CustomResource.new(crd: crd, **opts)
         else
+          type = definition["kind"]
           inst = new(**opts)
-          inst.type = definition["kind"]
+          inst.type = type
+          inst.global = true if global_names.map(&:downcase).include?(type.downcase)
```
Comment (KnVerey) [inline lib/kubernetes-deploy/kubernetes_resource.rb]:
```suggestion
          inst.global = global_names.map(&:downcase).include?(type.downcase)
```

File: lib/kubernetes-deploy/cluster_resource_discovery.rb [hunk 12]
```diff
@@ -16,8 +16,26 @@ def crds
       end
     end
 
+    def global_resource_names
```
Comment (KnVerey) [inline lib/kubernetes-deploy/cluster_resource_discovery.rb]: nit: this is a list of kinds, not names (which strictly speaking are defined separately)

File: lib/kubernetes-deploy/cluster_resource_discovery.rb [hunk 13]
```diff
@@ -16,8 +16,26 @@ def crds
...
+      @globals ||= fetch_globals.map { |g| g["kind"] }
+    end
+
     private
 
+    def fetch_globals
+      raw, _, st = kubectl.run("api-resources", "--namespaced=false", output: "wide", attempts: 5)
+      if st.success?
```
Comment (KnVerey) [inline lib/kubernetes-deploy/cluster_resource_discovery.rb]: I _think_ I understand what you're doing here, but this code is so dense that it is hard to follow. Can you please expand it to do fewer things per line. stick to descriptive variables, and maybe add a comment or two?

I'm also thinking this warrants a few unit tests showing it working with different column widths.

File: lib/kubernetes-deploy/cluster_resource_discovery.rb [hunk 14]
```diff
@@ -16,8 +16,26 @@ def crds
...
+        rows = raw.split("\n")
+        header = rows.shift.downcase.scan(/[a-z]+[\W]*/).each_with_object({}) do |match, hash|
+          start = hash.values.map(&:last).max.to_i
```
Comment (KnVerey) [inline lib/kubernetes-deploy/cluster_resource_discovery.rb]: This seems to be reflecting on the results of the block up to this point in the iteration. Could we be storing this value as we iterate instead of computing it like this? Isn't it the length of the header string we've processed so far, aka the end of the match we last processed?

File: test/unit/cluster_resource_discovery_test.rb [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+  def test_global_resource_kinds_failure
+    crd = mocked_cluster_resource_discovery(nil, success: false)
+    kinds = crd.global_resource_kinds
+    assert_equal(kinds, [])
+  end
+
+  def test_global_resource_kinds_success
+    crd = mocked_cluster_resource_discovery(full_response)
+    kinds = crd.global_resource_kinds
+    %w(MutatingWebhookConfiguration ComponentStatus ComponentStatus).each do |kind|
```
Comment (KnVerey) [inline test/unit/cluster_resource_discovery_test.rb]: The second and third are the same, presumably not intentional? I'd also suggest adding an assertion on the total number of kinds discovered.
